### DISCUSSION TITLE
Imports?

### DISCUSSION THREAD

**Discussion by prologic** [NicoNex/tau#30] [General]
Should we also add/build an import system?

For example the one I included in [monkey-lang](https://git.mills.io/prologic/monkey-lang#modules)

> Monkey supports modules. Modules are just like other Monkey source files with the extension .monkey. Modules are searched for > by SearchPaths which can be controlled by the environment MONKEYPATH. By default this is always the current directory.
> 
> To import a module:
>
> ``` 
> >> foo := import("foo")
> >> foo.A
> 5
> >> foo.Sum(2, 3)
> 5
> ```

**Comment by NicoNex** [comment]
The import system is what I'm currently working on and yes I took inspiration from your code, but the problem is that Tau has a different implementation so it needs to be adapted.
For the moment the import works for the eval and you can check it out in the [import2](https://github.com/NicoNex/tau/tree/import2) branch. 😃

**Comment by NicoNex** [comment]
Hi @prologic, the import is working correctly in the branch [import2](https://github.com/NicoNex/tau/tree/import2). Now what we should do is implement a working information hiding mechanism in the branch [information-hiding](https://github.com/NicoNex/tau/tree/information-hiding).

**Reply by prologic** [reply]
Can you elaborate on the "information hiding"?

**Reply by NicoNex** [reply]
Basically it's the exported/unexported mechanism.

**Reply by NicoNex** [reply]
The problem as of now is with the dot operator since it's handled in the same way inside modules and outside modules.

Due to the tree structure in the AST when a dot operation is executed it's impossible to know beforehand whether it's going to just fetch a value or set it. So I came up with an interface called `obj.GetSetter` that is implemented by an object called `obj.GetSetterImpl` like this:
```golang
type GetSetterImpl struct {
	GetFunc func() (Object, bool)
	SetFunc func(Object) Object
}
```

So when a dot is executed it will push on the stack an `obj.GetSetter` that acts in a transparent way and allows to do get and set operations later on.

For example in the VM:
```golang
func (vm *VM) execDot() error {
	var (
		right = vm.pop()
		left  = vm.pop()
	)

	switch l := left.(type) {
	case obj.MapGetSetter:
		return vm.push(&obj.GetSetterImpl{
			GetFunc: func() (obj.Object, bool) {
				return l.Get(right.String())
			},
			SetFunc: func(o obj.Object) obj.Object {
				return l.Set(right.String(), o)
			},
		})

	case obj.GetSetter:
		c := l.Object().(obj.Class)
		return vm.push(&obj.GetSetterImpl{
			GetFunc: func() (obj.Object, bool) {
				return c.Get(right.String())
			},

			SetFunc: func(o obj.Object) obj.Object {
				return c.Set(right.String(), o)
			},
		})

	default:
		return fmt.Errorf("%v object has no attribute %s", left.Type(), right)
	}
}
```

So for the exported/unexported mechanism we would need a way to differentiate when the dot operator is called inside a module or outside, and also it would be nice to make it work for objects too.

**Comment by NicoNex** [comment]
Hi @prologic, the import system is finally ready and working as expected.
The exported/unexported mechanism works both for entire modules and for the objects contained in them.

You can check it out both in the README and in the examples directory.
